theano/gpuarray/opt.py

```diff
@@ -250,53 +260,69 @@ def apply(self, fgraph):
-                mapping[i] = GpuFromHost(None)(i)
+                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
             else:
                 mapping[i] = i
-
-        # Iterating through output of all the nodes
-        for n in fgraph.toposort():
-            for o in n.outputs:
-                if isinstance(o.type, tensor.TensorType):
-                    mapping[o] = GpuFromHost(None)(o)
-                else:
-                    mapping[o] = o
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
 
         for node in fgraph.toposort():
 
-            # The Extra condition
-            if any([isinstance(i, GpuArrayVariable) or
-               isinstance(i, GpuArraySharedVariable)
-               for i in node.inputs + node.outputs]):
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
+                continue
 
-                move_to_GPU = False
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = False
+            if any([isinstance(i, GpuArrayVariable) or
+                    isinstance(i, GpuArraySharedVariable)
+                    for i in [mapping[v] for v in node.inputs] +
+                              node.outputs]):
 
-            # Oplifter's condition
-            # Will return a list of OP
-            # If None, means can't be moved.
+                move_to_GPU = True
 
             new_ops = None
-
-            # Selecting the best optimizer
-            # TODO : the tag should be updated to the one user provides
-            # currently using fast_run and fast_compile tag
+            # Apply the lifter
             for lopt in (self.local_optimizers_all +
                          self.local_optimizers_map.get(type(node.op), []) +
                          self.local_optimizers_map.get(node.op, [])):
-
-                new_ops = lopt.transform(node) or lopt(node)
-                break
-
-            if not new_ops or not isinstance(new_ops, theano.Op):
-                move_to_GPU = False
-
-            if move_to_GPU:
-                newnode = new_ops(*[mapping.get(i) for i in node.inputs])
-                for new_o, old_o in zip(newnode.outputs, node.outputs):
-                    mapping[old_o] = new_o
+                replace = False
+                for i in [mapping[i] for i in node.inputs]:
+                    if isinstance(i.type, GpuArrayType):
+                        context_name = i.type.context_name
+                        replace = True
+                        break
+                if replace:
+                    try:
+                        new_ops = lopt.transform(
+                            node, context_name,
+                            [mapping[i] for i in node.inputs])
+                    except TypeError:
+                        new_ops = lopt.transform(node, context_name)
```

Comment:
Change this try to:

                        new_ops = lopt.transform(
                            node.op, context_name,
                            [mapping[i] for i in node.inputs])